wnprc_billing/src/org/labkey/wnprc_billing/WNPRC_BillingController.java

```diff
@@ -472,59 +497,138 @@ public void setActivityID(String activityID)
+            PageFlowUtil.prepareResponseForFile(getViewContext().getResponse(), Collections.emptyMap(), pdfFile.getFileName(), invoicePdfForm.isAsAttachment());
+            pdfFile.getInvoicePDF().output(getViewContext().getResponse().getOutputStream());
+        }
+    }
 
-            double tierRate = accountTierRate != null ? accountTierRate.getTierRate() : 0;
-            String formName = invoicePdfForm.getName();
-            InvoicePDF pdf = formName.equals("Invoice PDF") ? new InvoicePDF(invoice, alias, invoiceRun, tierRate, contactEmail, billingAddress, creditToAccount) :
-                    new SummaryPDF(invoice, alias, invoiceRun, tierRate, contactEmail, billingAddress, creditToAccount);
+    private class PDFFile
+    {
+        String _fileName;
+        InvoicePDF _invoicePDF;
 
-            pdf.addPage();
-            List<InvoicedItem> invoicedItems;
+        public String getFileName()
+        {
+            return _fileName;
+        }
 
-            if (formName.equalsIgnoreCase("Invoice PDF"))
-            {
-                invoicedItems = getInvoicedItems(invoicePdfForm.getInvoiceNumber());
-                pdf.createLineItems(invoicedItems, true);
-            }
-            else
-            {
-                invoicedItems = getSummarizedItems(invoicePdfForm.getInvoiceNumber());
-                pdf.createLineItems(invoicedItems, false);
-            }
+        public void setFileName(String fileName)
+        {
+            _fileName = fileName;
+        }
 
-            SimpleDateFormat dateFormatBillingFor = new SimpleDateFormat("MM_yyyy");
-            String filename = alias.getGrantNumber() + "_" + dateFormatBillingFor.format(invoiceRun.getBillingPeriodStart()) + "_Invoice.pdf";
-            PageFlowUtil.prepareResponseForFile(getViewContext().getResponse(), Collections.emptyMap(), filename, invoicePdfForm.isAsAttachment());
-            pdf.output(getViewContext().getResponse().getOutputStream());
+        public InvoicePDF getInvoicePDF()
+        {
+            return _invoicePDF;
+        }
+
+        public void setInvoicePDF(InvoicePDF invoicePDF)
+        {
+            _invoicePDF = invoicePDF;
+        }
+    }
+
+    private PDFFile getInvoicePDF(String invoiceNumber, String formName) throws IOException
+    {
+        Invoice invoice = getInvoice(invoiceNumber);
+        if (null == invoice)
+        {
+            throw new NotFoundException(" Unable to generate PDF. The selected invoice could not be found.");
+        }
+        if (null == invoice.getInvoiceAmount())
+        {
+            throw new NotFoundException(" Unable to generate PDF. No Invoice Amount found.");
+        }
+
+        Alias alias = getInvoiceAccount(invoice.getAccountNumber());
+
+        if (null == alias)
+        {
+            throw new NotFoundException("Unable to generate PDF. Account Number not found.");
         }
+
+        InvoiceRun invoiceRun = getInvoiceRunByObjectId(invoice.getInvoiceRunId());
```

Comment:
Same security concern

Reply:
Filter on Container is now added, see the content of getInvoiceRunByObjectId().